rllib/core/rl_trainer/rl_trainer.py

```diff
@@ -173,12 +172,12 @@ def compute_loss(
             must contain one protected key "total_loss" which will be used for
             computing gradients through.
         """
-        # TODO: This method is built for multi-agent. While it is still possible to
-        # write single-agent losses, it may become confusing to users. We should find a
-        # way to allow them to specify single-agent losses as well, without having to
-        # think about one extra layer of hierarchy for module ids.
+        # TODO (Kourosh): This method is built for multi-agent. While it is still
+        # possible to write single-agent losses, it may become confusing to users. We
+        # should find a way to allow them to specify single-agent losses as well,
+        # without having to think about one extra layer of hierarchy for module ids.
```

Comment:
what if we have 2 functions:

1 called compute loss single agent, and 1 called compute loss multi agent. 

if you implement compute loss single agent, update calls that for all agents. If you call compute loss multi agent then update calls that function.

if you implement both we throw an error.

the one downside is I think it involves using the overrides decorator from ray/rllib

Reply:
I think we should roll a bit with multi-agent as first class citizen and see the impression of users. I have written BC trainer and honestly it's not that confusing. Especially if you think about the way people usually write these stuff for the first time, is that they most likely put breakpoints in the loss computation code to see what data they'll get and go from there. My hypothesis is that for an average "advanced" user (who is writing their own loss / algorithm) having the input as a MultiAgentSample batch is a good indicator of what they need to do plus they'll see examples of how other algorithm's losses are written. The advantage of this is that there is less api that a user would have to cope with hence a lower cognitive load.

Reply:
I have a proposal for this in a follow up PR.